### PR TITLE
[WIP] Porting Zephyr to Unisoc uwp5661 chip.

### PR DESCRIPTION
**Pull request by easonxiang**
Unisoc UWP5661 is a highly integrated connectivity single chip which offers the lowest BOM in the industry for smart phone, PC, STB, OTT, IoT and automotive applications.
This chip includes 2.4GHz and 5GHz WLAN IEEE 802.11 a/b/g/n/ac, Bluetooth 5.0 with supporting high power mode, Direction Finding and Long Range.

This patch is initial commit, which include uwp5661 soc support, hal library, watchdog and uart driver.

### CODE REVIEW

**Comment by galak** [comment]
Thanks for the submission.  Please review http://docs.zephyrproject.org/contribute/contribute_guidelines.html  and http://docs.zephyrproject.org/contribute/contribute_non-apache.html

1. Commit messages are required in each commit
2. Imports to EXT need a clear license statement and README.

Thanks

**Comment by galak** [comment]
> @galak Thanks for the review, patches have been updated as requirement. EXT files are under apache2.0 license. Please comment if hvae any problem.
> 
> Thanks.

Can you update all src files to have a SPDX tag in them to be clear on the license
```
/*
 * Copyright (c) 2018 Linaro Limited.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
```

**Comment by galak** [comment]
> @galak Thanks for the review, patches have been updated as requirement. EXT files are under apache2.0 license. Please comment if hvae any problem.

How does one get the HAL/SDK for the Unisoc UWP360?

**Comment by easonxiang** [comment]
Unisoc UWP360 HAL/SDK is under development, more module interfaces/drivers will be added after various tests. The full SDK will be released once it was ready.

File: arch/arm/soc/unisoc_uwp/uwp3/Kconfig.series [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Kconfig.series - UNISOC UWP3
+#
+
+config  SOC_SERIES_UWP3
+	bool "Unisoc UWP3"
+	select CPU_CORTEX_M
```
Comment (ioannisg) [inline arch/arm/soc/unisoc_uwp/uwp3/Kconfig.series]: Not needed to specify CPU_CORTEX_M, if you have already selected CPU_CORTEX_M4

**Review by ioannisg — CHANGES_REQUESTED** [review]

File: arch/arm/soc/unisoc_uwp/uwp3/Kconfig.series (hunk 1 above)
Comment (easonxiang) [reply arch/arm/soc/unisoc_uwp/uwp3/Kconfig.series]: Thanks for the review, patch has been updated.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
A board .rst file that includes information like where to get such a board must be included.

**Comment by galak** [comment]
@easonxiang any updates on the availability of the HAL/SDK?  and on updating this PR.

**Comment by easonxiang** [comment]
@SebastianBoe The Soc Chip and Board are not announced yet, the document for zephyr is no available currently, I will update it when it was ready.

**Comment by galak** [comment]
Any updates on the availability of the SDK/HAL from a public website?

**Comment by easonxiang** [comment]
There is not. Actually, Github is our first commit to the public website.

Currently, we are debugging Wi-Fi functionality, scan and connect are works on Zephyr system.

**Comment by pfalcon** [comment]
@ioannisg : As your comment seems to have been addressed, perhaps you could dismiss your -1.

**Comment by pfalcon** [comment]
@easonxiang : This now has conflicts, can you please rebase on master?

**Comment by easonxiang** [comment]
@pfalcon Thanks for your comment, I wanna to know if it is better to rebase on 1.13 release as  there are so many changes on master and our patches?

**Comment by easonxiang** [comment]
In addition, we are preparing a demo for the coming Linaro Connect, so we may rebase on the latest code after the Connect.

**Comment by pfalcon** [comment]
@easonxiang:

> I wanna to know if it is better to rebase on 1.13 release as there are so many changes on master and our patches?

I would suggest to rebase regularly, e.g. at least once a week. This way you won't accumulate conflicts and discrepancies with mainline, but resolve them quickly one by one as they appear. Of course, that might not work for every workflow, so up to you.

> we are preparing a demo for the coming Linaro Connect

Cool, looking forward to it.

**Comment by galak** [comment]
Can we rebase this PR.  Also I understand that the SoC/chip got renamed, so we should probably make that change as well.

**Comment by easonxiang** [comment]
@galak Thanks for your comment.
This PR has been rebased on the latest code, and chip name was renamed to uwp5661.

File: boards/arm/96b_ivy5661/doc/96b_ivy5661.rst [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     :width: 400px
+     :align: center
+     :height: 600px
+     :alt: 96Boards IVY5661
+
+     96Boards IVY5661
+
+This board support Wi-Fi 802.11ac 2x2 dual band and Bluetooth 5 with
+high power mode. It can be used by intelligent household, IP camera,
+Wi-Fi repeater and other networking applications.
```
Comment (dbkinder) [inline boards/arm/96b_ivy5661/doc/96b_ivy5661.rst]: add a comma:  ``repeater, and``

**Review by dbkinder — CHANGES_REQUESTED** [review]
doc look fine, with a couple of little tweaks please.

File: boards/arm/96b_ivy5661/doc/96b_ivy5661.rst [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
++-----------+------------+-------------------------------------+
+| PINMUX    | on-chip    | pinmux                              |
++-----------+------------+-------------------------------------+
+| FLASH     | on-chip    | flash                               |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c                                 |
++-----------+------------+-------------------------------------+
+
+
+The default board configuration can be found in the defconfig file:
```
[the remark this answers is not in the document]
Comment (easonxiang) [reply boards/arm/96b_ivy5661/doc/96b_ivy5661.rst]: >This renders on the website better as:
>
>The default board configuration can be found in the defconfig file:
>``boards/arm/96b_ivy5661/96b_ivy5661_defconfig``

Thanks for the review.
Do you mean remove the blank line between these two lines?

File: boards/arm/96b_ivy5661/doc/96b_ivy5661.rst (hunk 2 above)
Comment (easonxiang) [reply boards/arm/96b_ivy5661/doc/96b_ivy5661.rst]: Done.

File: boards/arm/96b_ivy5661/doc/96b_ivy5661.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (easonxiang) [reply boards/arm/96b_ivy5661/doc/96b_ivy5661.rst]: done.

File: boards/arm/96b_ivy5661/board.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+/* Push button switch 2 */
+#define SW2_GPIO_PIN	6    /* GPIO22/Pin15 */
```
Comment (galak) [inline boards/arm/96b_ivy5661/board.h]: We should move this into the device tree, take a look at this commit to see an example:

commit 619049a74c8b7179ca4925f243597359e4930640
Author: Kumar Gala <kumar.gala@linaro.org>
Date:   Fri Sep 28 16:35:16 2018 -0500

    boards: nrf51_ble400: Convert to using dts for led/button

File: boards/arm/96b_ivy5661/dbghdr.c [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifdef BOARD_96B_IVY5661
+__attribute__ ((section(".dbghdr")))
+const unsigned long ulDebugHeader[] = {
```
Comment (galak) [inline boards/arm/96b_ivy5661/dbghdr.c]: Is this a board specific header, or something we'd need for this SoC always?

File: boards/arm/96b_ivy5661/support/openocd.cfg [hunk 6]
```diff
@@ -0,0 +1,12 @@
+source [find board/stm32f4discovery.cfg]
```
Comment (galak) [inline boards/arm/96b_ivy5661/support/openocd.cfg]: Is this correct, saying the board cfg is stm32f4discovery.cfg seems wrong

File: drivers/serial/uart_uwp.c [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <arch/cpu.h>
+#include <uart.h>
+
+#include "uwp_hal.h"
+#include "uart_uwp.h"
+#include "../interrupt_controller/intc_uwp.h"
```
Comment (galak) [inline drivers/serial/uart_uwp.c]: Is this header needed?

File: drivers/serial/uart_uwp.c [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+	return (status & (UART_TXF_EMPTY | UART_RXF_FULL));
+}
+
+static int uart_uwp_irq_update(struct device *dev)
+{
+	return 1;
+}
+
+
+	void (*irq_callback_set)(struct device *dev,
```
Comment (galak) [inline drivers/serial/uart_uwp.c]: Why do we need this?

File: dts/arm/unisoc/mem.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2018, UNISOC Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DT_UNISOC_MEM_H
+#define __DT_UNISOC_MEM_H
+
+#define __SIZE_K(x) (x * 1024)
```
Comment (galak) [inline dts/arm/unisoc/mem.h]: We've moved away from this style of mem.h header.  We introduced DT_SIZE_K/DT_SIZE_M.

**Review by galak — CHANGES_REQUESTED** [review]
Various cleanups/comments.

File: dts/arm/unisoc/mem.h (hunk 9 above)
Comment (easonxiang) [reply dts/arm/unisoc/mem.h]: done.

File: drivers/serial/uart_uwp.c (hunk 8 above)
Comment (easonxiang) [reply drivers/serial/uart_uwp.c]: This is a mistake, remove it.

File: boards/arm/96b_ivy5661/dbghdr.c (hunk 5 above)
Comment (easonxiang) [reply boards/arm/96b_ivy5661/dbghdr.c]: This is not used by the SoC, remove this file.

File: drivers/serial/uart_uwp.c (hunk 7 above)
Comment (easonxiang) [reply drivers/serial/uart_uwp.c]: This header file is needed by uart driver.
There is a 2-Level interrupt mechanism in this chip. UART2 interrupt line connect to interrupt controller, and interrupt controller connect to NVIC. Therefor, interrupt controller need to provide interfaces to register uart isr callback, such as uwp_aon_intc_set_irq_callback.

File: boards/arm/96b_ivy5661/board.h (hunk 4 above)
Comment (easonxiang) [reply boards/arm/96b_ivy5661/board.h]: This modification will take some times.

File: boards/arm/96b_ivy5661/support/openocd.cfg (hunk 6 above)
Comment (easonxiang) [reply boards/arm/96b_ivy5661/support/openocd.cfg]: Yes, this it not correct. 
UWP5661 does just support J-Link for debugging. We are still looking for the root cause that how to support OpenOCD as well.
May I remove this config for this commit, and support OpenOCD in the future.

**Comment by easonxiang** [comment]
Split this patch into 2 parts:

Unisoc HAL, imported from https://github.com/unisoc/hal_uwp5
https://github.com/zephyrproject-rtos/zephyr/pull/10622 

and ivy5661 board support:
https://github.com/zephyrproject-rtos/zephyr/pull/10626

**Pull request closed without merging**